q2cli/core/usage.py

```diff
@@ -0,0 +1,146 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016-2020, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from qiime2.core.type.primitive import Bool
+
+import qiime2.sdk.usage as usage
+from qiime2.sdk.util import (
+    is_metadata_type,
+    is_metadata_column_type,
+    is_visualization_type,
+)
+
+from q2cli.util import to_cli_name
+from q2cli.util import to_snake_case
+
+
+class CLIUsage(usage.Usage):
+    def __init__(self):
+        super().__init__()
+        self._recorder = []
+        self._init_data_refs = dict()
+        self._metadata_refs = dict()
+        self._col_refs = dict()
+        self._merge_targets = []
+
+    def _init_data_(self, ref, factory):
+        self._init_data_refs[ref] = factory
+        # Don't need to compute anything, so just pass along the ref
+        return ref
+
+    def _merge_metadata_(self, ref, records):
+        merge_target = ref
+        self._merge_targets.append(merge_target)
+        for record in records:
+            self._metadata_refs[record.ref] = record.result
+        return merge_target
+
+    def _get_metadata_column_(self, ref, record, column_name):
+        self._metadata_refs[record.ref] = record.result
+        self._col_refs[record.ref] = column_name
+        return record.result
+
+    def _comment_(self, text: str):
+        self._recorder.append('# %s' % (text,))
+
+    def _action_(
+        self, action: usage.UsageAction, input_opts: dict, output_opts: dict
+    ):
+        action_f, action_sig = action.get_action()
+        inputs, params = self.extract(action_f, input_opts)
+        t = self._template_action(action_f, inputs, params, output_opts)
+        self._recorder.append(t)
+        return output_opts
```

Comment:
Nice! This is coming along well! To elaborate a bit more on my request in the last round of review, I'm thinking something like this:

```suggestion
    def _action_(self, action, input_opts, output_opts):
        action_f, action_sig = action.get_action()
        params, mds = self._extract_from_signature(action_f, input_opts)
        t = self._template_action(action_f, input_opts, params, mds, output_opts)
        self._recorder.append(t)
        return output_opts
        
        ...
        
    def _extract_from_signature(action_f, input_opts):
        ...
        return (extracted_params, extracted_mds)
        
    def _template_action(action_f, inputs, params, mds, outputs):
        inputs_t = self._template_inputs(action_f, inputs)
        params_t = self._template_parameters(action_f, params)
        mds_t = self._template_metadata(action_f, mds)
        outputs_t = self._template_outputs(action_f, outputs)
        
        cmd = to_cli_name(f"qiime {action_f.plugin_id} {action_f.id}")
        t = " \\\n".join([cmd] + inputs_t + params_t + mds_t + outputs_t)
        return t
```

I think with a structure like this in place you might be able to reduce the need for the state attributes tacked onto the class (but we can talk about that more in the next round).

So, rephrasing the thought above, we might see an API like this:

```python
class CLIUsage(usage.Usage):
    # from parent class
    def __init__(self): pass
    def _init_data_(self, ref, factory): pass
    def _merge_metadata_(self, ref, records): pass
    def _get_metadata_column_(self, ref, record, column_name): pass
    def _comment_(self, text: str): pass
    def _action_(self, action, input_opts, output_opts): pass
    def _assert_has_line_matching_(self, ref, label, path, expression): pass
    # for this child class
    ## private api
    def _extract_from_signature(self, action_f, input_opts): pass
    def _template_action(self, action_f, inputs, params, mds, outputs): pass
    def _template_inputs(self, action_f, inputs): pass
    def _template_parameters(self, action_f, params): pass
    def _template_metadata(self, action_f, mds): pass
    def _template_outputs(self, action_f, outputs): pass
    ## public api
    def render(self): pass
    def get_example_data(self): pass
```